rasterio/_shim21.pyx

```diff
@@ -11,6 +15,53 @@ include "shim_rasterioex.pxi"
 cdef extern from "gdal.h" nogil:
 
     cdef CPLErr GDALDeleteRasterNoDataValue(GDALRasterBandH hBand)
+    GDALDatasetH GDALOpenEx(const char *filename, int flags, const char **allowed_drivers, const char **options, const char **siblings) # except -1
+
+from rasterio._err cimport exc_wrap_pointer
+
+
+cdef GDALDatasetH open_dataset(
+        object filename, int flags, object allowed_drivers, object open_options,
+        object siblings) except NULL:
+    """Wrapper for GDALOpen and GDALOpenShared"""
+    cdef const char *fname = NULL
+    cdef const char **drivers = NULL
+    cdef const char **options = NULL
+    cdef const char *key = NULL
+    cdef const char *val = NULL
+    cdef const char *driver = NULL
+    cdef GDALDatasetH hds = NULL
+
+    filename = filename.encode('utf-8')
+    fname = filename
+
+    # Construct a null terminated C list of driver
+    # names for GDALOpenEx.
+    if allowed_drivers:
+        for name in allowed_drivers:
+            name = name.encode('utf-8')
+            driver = name
+            drivers = CSLAddString(drivers, driver)
+
+    for k, v in open_options.items():
+        k = k.upper().encode('utf-8')
+        key = k
+
+        # Normalize values consistent with code in _env module.
+        if isinstance(v, bool):
+            v = ('ON' if v else 'OFF').encode('utf-8')
```

Comment:
I'm concerned about our normalization of GDAL config option names and I think its time to encapsulate this behavior in a normalization function to help us be more consistent.  This PR introduces a [check for `BLOCKXSIZE` and `BLOCKYSIZE`](https://github.com/mapbox/rasterio/pull/1196/files#diff-49692333e8667a6b046cbdcad3c1164cR931) that should be passed through a similar function.  The overall problem is better handled in https://github.com/mapbox/rasterio/issues/1149 and https://github.com/mapbox/rasterio/issues/1012.  Ensuring we have consistent config option names and values is critical for environment discovery on GDAL versions lacking `CPLGetConfigOptions()`.

Reply:
Yes, I'm with you on normalization, but as these are creation and open options, they don't affect the global environment.

Reply:
Our creation option normalization is not correct in all cases.
 as documented in https://github.com/mapbox/rasterio/issues/1149.  Passing everything through a single function would allow us to catch special cases and other GDAL oddities.  I suspect the real solution is to ask that GDAL support `TRUE` and `YES` for enabling and `FALSE` and `NO` for disabling for included drivers.